### PR TITLE
Upgrade controller-runtime to 0.18.1

### LINKED ISSUE #7766: Support controller-runtime v0.18.0
https://github.com/kubernetes-sigs/controller-runtime/releases/tag/v0.18.0

Unfortunately there are quite a few breaking changes in v0.18.0, specifically generics support. See https://github.com/kubernetes-sigs/controller-runtime/pull/2783

relates: https://github.com/elastic/cloud-on-k8s/pull/7733

### PR DESCRIPTION
**Pull request by naemono**
Fixes #7766

This upgrades the controller-runtime to 0.18.1, which is a larger issue than normal as generics are introduced to the code base.

## Notes for review
- [x] lint issues will be resolved with https://github.com/elastic/ci-agent-images/pull/659
- [x] still working through the other issues with `make generate`.

### CODE REVIEW

**Review by pebrc — COMMENTED** [review]
Most of the type params you added are unnecessary. I pushed a commit to illustrate, there are a few more controllers to clean up that I did not do.

File: pkg/controller/beat/auditbeat/auditbeat.go [hunk 1]
```diff
@@ -15,16 +17,16 @@ const (
 	Type beatcommon.Type = "auditbeat"
 )
 
-type Driver struct {
+type Driver[T client.Object] struct {
```
Comment (pebrc) [inline pkg/controller/beat/auditbeat/auditbeat.go:20]: unnecessary

File: pkg/controller/beat/heartbeat/heartbeat.go [hunk 2]
```diff
@@ -15,15 +17,15 @@ const (
 	Type beatcommon.Type = "heartbeat"
 )
 
-type Driver struct {
+type Driver[T client.Object] struct {
```
Comment (pebrc) [inline pkg/controller/beat/heartbeat/heartbeat.go:20]: unecessary

File: pkg/controller/beat/journalbeat/journalbeat.go [hunk 3]
```diff
@@ -15,15 +17,15 @@ const (
 	Type beatcommon.Type = "journalbeat"
 )
 
-type Driver struct {
+type Driver[T client.Object] struct {
```
Comment (pebrc) [inline pkg/controller/beat/journalbeat/journalbeat.go:20]: unnecessary

File: pkg/controller/beat/metricbeat/metricbeat.go [hunk 4]
```diff
@@ -15,16 +17,16 @@ const (
 	Type beatcommon.Type = "metricbeat"
 )
 
-type Driver struct {
+type Driver[T client.Object] struct {
```
Comment (pebrc) [inline pkg/controller/beat/metricbeat/metricbeat.go:20]: unnecessary

File: pkg/controller/beat/otherbeat/otherbeat.go [hunk 5]
```diff
@@ -5,20 +5,22 @@
...
 import (
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
 	beatv1beta1 "github.com/elastic/cloud-on-k8s/v2/pkg/apis/beat/v1beta1"
 	beatcommon "github.com/elastic/cloud-on-k8s/v2/pkg/controller/beat/common"
 	"github.com/elastic/cloud-on-k8s/v2/pkg/controller/common/reconciler"
 )
 
-type Driver struct {
+type Driver[T client.Object] struct {
```
Comment (pebrc) [inline pkg/controller/beat/otherbeat/otherbeat.go:15]: unnecessary

File: pkg/controller/beat/packetbeat/packetbeat.go [hunk 6]
```diff
@@ -15,16 +17,16 @@ const (
 	Type beatcommon.Type = "packetbeat"
 )
 
-type Driver struct {
+type Driver[T client.Object] struct {
```
Comment (pebrc) [inline pkg/controller/beat/packetbeat/packetbeat.go:20]: unnecessary

File: pkg/controller/common/watches/pods.go [hunk 7]
```diff
@@ -18,16 +18,11 @@ import (
 )
 
 func Add(mgr manager.Manager, c controller.Controller, owner client.Object, objects ...client.Object) error {
```
Comment (pebrc) [inline pkg/controller/common/watches/pods.go:20]: Is this function used?

File: pkg/controller/kibana/controller.go [hunk 8]
```diff
@@ -63,17 +64,17 @@ func newReconciler(mgr manager.Manager, params operator.Parameters) *ReconcileKi
 	}
 }
 
-func addWatches(mgr manager.Manager, c controller.Controller, r *ReconcileKibana) error {
+func addWatches[T client.Object](mgr manager.Manager, c controller.Controller, r *ReconcileKibana) error {
```
Comment (pebrc) [inline pkg/controller/kibana/controller.go:67]: That type parameter seems unnecessary too maybe?

File: pkg/controller/remoteca/watches.go [hunk 9]
```diff
@@ -57,8 +58,8 @@ func addWatches(mgr manager.Manager, c controller.Controller, r *ReconcileRemote
 
 // newRequestsFromMatchedLabels creates a watch handler function that creates reconcile requests based on the
 // labels set on a Secret which contains the remote CA.
-func newRequestsFromMatchedLabels() handler.MapFunc {
-	return func(ctx context.Context, obj client.Object) []reconcile.Request {
+func newRequestsFromMatchedLabels[T client.Object]() handler.TypedMapFunc[T] {
```
Comment (pebrc) [inline pkg/controller/remoteca/watches.go:61]: Unecesssary parameterisation, can be instantiated directly to `corev1.Secret`

**Review by pebrc — CHANGES_REQUESTED** [review]

File: pkg/controller/common/watches/pods.go (hunk 7 above)
Comment (naemono) [reply pkg/controller/common/watches/pods.go]: Doesn't seem to be. I've removed it.

**Comment by naemono** [comment]
> Most of the type params you added are unnecessary. I pushed a commit to illustrate, there are a few more controllers to clean up that I did not do.

@pebrc this is ready for the next set of 👀 . Thanks much for the initial push in the right direction.

File: pkg/controller/beat/auditbeat/auditbeat.go (hunk 1 above)
Comment (pebrc) [reply pkg/controller/beat/auditbeat/auditbeat.go:20]: Did you forget to push a commit maybe?

**Review by pebrc — DISMISSED** [review]
LGTM code-wise, let's make sure we run a full test-suite before merging this.

**Comment by naemono** [comment]
> LGTM code-wise, let's make sure we run a full test-suite before merging this.

Will do. 

Also, as you noted elsewhere, there's likely a further simplification since `type EventHandler TypedEventHandler[client.Object]` exists, so we likely don't have to use `TypedEventHandler` at all. Let me go down that path a bit and see if it further simplifies this...

**Comment by naemono** [comment]
>  Let me go down that path a bit and see if it further simplifies this...

Never mind on this. We discussed externally and feel that the controller-runtime is heading in this direction (using Typed*), so we'll leave it as is.

**Comment by naemono** [comment]
Finally got the error as my local environment was having some major issues:
```
2024-05-06T14:09:29.021-0500	ERROR	controller-runtime.source.EventHandler	failed to get informer from cache	{"service.version": "2.14.0-SNAPSHOT+08c73c3d", "error": "Timeout: failed waiting for *v1.Pod Informer to sync"}
sigs.k8s.io/controller-runtime/pkg/internal/source.(*Kind[...]).Start.func1.1
	/Users/mmontgomery/go/1.21.6/pkg/mod/sigs.k8s.io/controller-runtime@v0.18.0/pkg/internal/source/kind.go:76
k8s.io/apimachinery/pkg/util/wait.loopConditionUntilContext.func1
	/Users/mmontgomery/go/1.21.6/pkg/mod/k8s.io/apimachinery@v0.30.0/pkg/util/wait/loop.go:53
k8s.io/apimachinery/pkg/util/wait.loopConditionUntilContext
	/Users/mmontgomery/go/1.21.6/pkg/mod/k8s.io/apimachinery@v0.30.0/pkg/util/wait/loop.go:54
k8s.io/apimachinery/pkg/util/wait.PollUntilContextCancel
	/Users/mmontgomery/go/1.21.6/pkg/mod/k8s.io/apimachinery@v0.30.0/pkg/util/wait/poll.go:33
sigs.k8s.io/controller-runtime/pkg/internal/source.(*Kind[...]).Start.func1
	/Users/mmontgomery/go/1.21.6/pkg/mod/sigs.k8s.io/controller-runtime@v0.18.0/pkg/internal/source/kind.go:64
2024-05-06T14:09:29.021-0500	ERROR	manager	Operator stopped with error	{"service.version": "2.14.0-SNAPSHOT+08c73c3d", "error": "must specify Channel.Handler"}
github.com/elastic/cloud-on-k8s/v2/cmd/manager.doRun.func2
	/Users/mmontgomery/git/cloud-on-k8s/cmd/manager/main.go:403
```

Investigating....

**Comment by pebrc** [comment]
@naemono I have a hunch things will work if we go back to `client.Object` and avoid the `Typed*` handlers:

```golang
	// Watch for changes to Elasticsearch
	if err := c.Watch(
		source.Kind(mgr.GetCache(), client.Object(&esv1.Elasticsearch{}), &handler.EnqueueRequestForObject{})); err != nil {
		return err
	}
```

**Comment by pebrc** [comment]
@naemono ignore my previous comment: 

```
func WatchClusterHealthChange(m *Manager) source.Source {
	evtChan := make(chan event.TypedGenericEvent[*esv1.Elasticsearch])
	m.AddObservationListener(healthChangeListener(evtChan))
	// Each event in Source will be consumed and turned into
	// a reconciliation request.
	//
	// DestBufferSize is kept at the default value (1024).
	// This means we can enqueue a maximum of 1024 requests
	// before blocking observers from moving on.
	return source.Channel(evtChan, nil)
}
```
I think this is problem.

**Comment by naemono** [comment]
> @naemono ignore my previous comment:
> 
> ```
> func WatchClusterHealthChange(m *Manager) source.Source {
> 	evtChan := make(chan event.TypedGenericEvent[*esv1.Elasticsearch])
> 	m.AddObservationListener(healthChangeListener(evtChan))
> 	// Each event in Source will be consumed and turned into
> 	// a reconciliation request.
> 	//
> 	// DestBufferSize is kept at the default value (1024).
> 	// This means we can enqueue a maximum of 1024 requests
> 	// before blocking observers from moving on.
> 	return source.Channel(evtChan, nil)
> }
> ```
> 
> I think this is problem.

Thanks. I'll take a look here. I've been digging through the runtime code yesterday to try to see what we were missing in regards to `must specify Channel.Handler` but I felt like I wasn't making huge progress.

**Comment by naemono** [comment]
> source.Channel(evtChan, nil)

Yeah, now that you've pointed this out, it seems quite a bit more obvious. I'm testing the fix now. Thanks again.

**Comment by naemono** [comment]
> > source.Channel(evtChan, nil)
> 
> Yeah, now that you've pointed this out, it seems quite a bit more obvious. I'm testing the fix now. Thanks again.

[81666c9](https://github.com/elastic/cloud-on-k8s/pull/7768/commits/81666c9ed56ae69cfb25a08c0e432ac6b58cf999) seems to fix it locally for me.

Also note I upgraded to 0.18.1 because of some bug fixes that were released shortly after .0. https://github.com/kubernetes-sigs/controller-runtime/releases/tag/v0.18.1

**Comment by naemono** [comment]
E2e Smoke test succeeded now. Let me know if you think further e2e test runs are worthwhile here.

**Comment by thbkrkr** [comment]
buildkite test this -f p=kind,s=8.12.2 -m t=TestAPMServerVersionUpgradeToLatest8x,t=TestAgentVersionUpgradeToLatest8x,t=TestElasticMapsServerVersionUpgradeToLatest8x,t=TestEnterpriseSearchVersionUpgradeToLatest8x,t=TestFleetKubernetesNonRootIntegrationRecipe

File: pkg/controller/autoscaling/elasticsearch.go [hunk 10]
```diff
@@ -27,8 +28,8 @@ func Add(mgr manager.Manager, p operator.Parameters) error {
 	if err != nil {
 		return err
 	}
-	if err := controller.Watch(source.Kind(mgr.GetCache(), &v1alpha1.ElasticsearchAutoscaler{}), &handler.EnqueueRequestForObject{}); err != nil {
+	if err := controller.Watch(source.Kind(mgr.GetCache(), &v1alpha1.ElasticsearchAutoscaler{}, &handler.TypedEnqueueRequestForObject[*v1alpha1.ElasticsearchAutoscaler]{})); err != nil {
 		return err
 	}
-	return controller.Watch(source.Kind(mgr.GetCache(), &esv1.Elasticsearch{}), reconciler.Watches.ReferencedResources)
+	return controller.Watch(source.Kind(mgr.GetCache(), (client.Object)(&esv1.Elasticsearch{}), reconciler.Watches.ReferencedResources))
```
Comment (thbkrkr) [inline pkg/controller/autoscaling/elasticsearch.go:34]:
```suggestion
	return controller.Watch(source.Kind[client.Object](mgr.GetCache(), &esv1.Elasticsearch{}, reconciler.Watches.ReferencedResources))
```

**Review by thbkrkr — APPROVED** [review]
LGTM

Thank you @naemono for taking care of this.

**Comment by pebrc** [comment]
I think there is 0.18.2 already btw.

**Review by pebrc — APPROVED** [review]

**Comment by naemono** [comment]
> I think there is 0.18.2 already btw.

lol. When I looked yesterday morning, there wasn't. I'll look and see if it's relevant for our operator and update if appropriate.

**Comment by naemono** [comment]
> > I think there is 0.18.2 already btw.
> 
> lol. When I looked yesterday morning, there wasn't. I'll look and see if it's relevant for our operator and update if appropriate.

It's not used by us specifically, but it could be used for use to avoid having to enable our own http server when dev mode is enabled, so might as well upgrade now.

**Pull request merged**